consensus/hotstuff/validator/validator.go

```diff
@@ -12,89 +12,143 @@ import (
 
 // Validator is responsible for validating QC, Block and Vote
 type Validator struct {
-	// TODO: change to hotstuff.Replicas when front-loading QC verification
-	// https://github.com/onflow/flow-go/pull/2328#discussion_r866494368
-	committee hotstuff.DynamicCommittee
-	forks     hotstuff.ForksReader
+	committee hotstuff.Replicas
 	verifier  hotstuff.Verifier
 }
 
 var _ hotstuff.Validator = (*Validator)(nil)
 
 // New creates a new Validator instance
 func New(
-	committee hotstuff.DynamicCommittee,
-	forks hotstuff.ForksReader,
+	committee hotstuff.Replicas,
 	verifier hotstuff.Verifier,
 ) *Validator {
 	return &Validator{
 		committee: committee,
-		forks:     forks,
 		verifier:  verifier,
 	}
 }
 
-// ValidateQC checks the validity of a QC for a given block. Inputs:
-//  * qc - the qc to be validated
-//  * block - the block that the qc is pointing to
+// ValidateTC validates the TC
+// tc - the tc to be validated
 // During normal operations, the following error returns are expected:
-//  * model.InvalidBlockError if the QC is invalid
-func (v *Validator) ValidateQC(qc *flow.QuorumCertificate, block *model.Block) error {
-	if qc.BlockID != block.BlockID {
-		// Sanity check! Failing indicates a bug in the higher-level logic
-		return fmt.Errorf("qc.BlockID %s doesn't match block's ID %s", qc.BlockID, block.BlockID)
+//  * model.InvalidTCError if the TC is invalid
+//  * model.ErrViewForUnknownEpoch if the TC refers unknown epoch
+// Any other error should be treated as exception
+func (v *Validator) ValidateTC(tc *flow.TimeoutCertificate) error {
+	highestQC := tc.TOHighestQC
+
+	// there is no reason for a timeout certificate to include a QC for the same view
+	// (if you have a QC for the view, we can always use the QC directly)
+	// however, since QCs and TCs are processed/created asynchronously, we allow this case
+	if tc.View < highestQC.View {
+		return newInvalidTCError(tc, fmt.Errorf("TC's QC cannot be newer than the TC's view"))
+	}
+
+	highestQCView := tc.TOHighQCViews[0]
+	for _, view := range tc.TOHighQCViews {
+		if highestQCView < view {
+			highestQCView = view
+		}
+	}
+
+	if highestQCView != tc.TOHighestQC.View {
+		return newInvalidTCError(tc, fmt.Errorf("included QC (view=%d) should be equal to highest contributed view: %d", tc.TOHighestQC.View, highestQCView))
+	}
+
+	// 1. Check if there is super-majority of votes
+	allParticipants, err := v.committee.IdentitiesByEpoch(tc.View)
+	if err != nil {
+		return fmt.Errorf("could not get consensus participants at view %d: %w", tc.View, err)
+	}
+
+	signers, err := signature.DecodeSignerIndicesToIdentities(allParticipants, tc.SignerIndices)
+	if err != nil {
+		if signature.IsDecodeSignerIndicesError(err) {
```

Comment:
I specifically asked for this method to be removed, because I consider it an intrinsically unsafe pattern (see reasoning in description of [this PR](https://github.com/onflow/flow-go/pull/2517)). 
In master, `IsDecodeSignerIndicesError` had been replaced by `InvalidSignerIndicesError`.

Reply:
we can clean this up in a later PR. Just wanted to highlight the change.